sdk/go/common/resource/testing/rapid.go

```diff
@@ -125,32 +237,65 @@ func ArchivePropertyGenerator(maxDepth int) *rapid.Generator {
 
 // ResourceReferenceGenerator generates resource.ResourceReference values.
 func ResourceReferenceGenerator() *rapid.Generator {
+	return resourceReferenceGenerator(nil)
+}
+
+func resourceReferenceGenerator(ctx *StackContext) *rapid.Generator {
+	var resourceGenerator *rapid.Generator
+	if ctx == nil {
+		resourceGenerator = rapid.Custom(func(t *rapid.T) *resource.State {
+			id := resource.ID("")
+			custom := !rapid.Bool().Draw(t, "component").(bool)
+			if custom {
+				id = IDGenerator().Draw(t, "resource ID").(resource.ID)
+			}
+
+			return &resource.State{
+				URN:    URNGenerator().Draw(t, "resource URN").(resource.URN),
+				Custom: custom,
+				ID:     id,
+			}
+		})
+	} else {
+		resourceGenerator = rapid.SampledFrom(ctx.Resources())
+	}
+
 	return rapid.Custom(func(t *rapid.T) resource.ResourceReference {
-		id := rapid.OneOf(UnknownPropertyGenerator(), StringPropertyGenerator()).Draw(t, "referenced ID").(resource.PropertyValue)
-		if id.IsString() && id.StringValue() == "" {
-			id = resource.PropertyValue{}
+		r := resourceGenerator.Draw(t, "referenced resource").(*resource.State)
+
+		var id resource.PropertyValue
+		if r.Custom {
```

Comment:
If r.Custom == false, we get id set to default empty value? Why is that? Feels a bit odd, perhaps worth a quick comment here.

Reply:
This is because component resources do not have IDs, and we use the absence of a value in the `ID` field to indicate that.